Added souper post (#421)
mono/website: _posts/2018-11-21-souper.md

Other projects may want to get a feel for how fast they could run
if they were to invest making sure mono generates good code.
If the compile time increase doesn't discourage them, many
projects may find some benefit in such an optimizing compiler.

I recommend that curious readers install Z3, get a checkout of

[https://github.com/google/souper](https://github.com/google/souper),

and complete the compilation process described in that documentation.

When AOTing code with mono, they're going to want to pass the
commandline flags named there into the `---aot=llvmopts=` argument.

As of the time of this writing, that is

```
llvmopts="-load /path/to/libsouperPass.so -souper -z3-path=/usr/bin/z3" 
``` 

Mono will then allow Souper to step in during the middle of the LLVM
compilation and try it's best at brute-forcing some better code. If
there's anything short and fast that does the job better, it will be
found.

It is frankly amazing that Mono can get such extensive optimizations simply
by compiling to LLVM IR. Without changing a single line of Mono's source, we
changed our compilation pipeline in truly dramatic ways. This speaks to Mono's

> flexibility as an embedded runtime. Embedders using mono should consider

Reviewer: > This speaks to Mono's flexibility as an embedded runtime.

Does it?  Feels more like it speaks to LLVM's flexibility as a compiler framework.

> using our LLVM backend with these new community passes. Feedback about the

Reviewer: what are "new community passes"?

---
layout: blog
title: Giving Mono Souper Powers
author: Alexander Kyte
tags: [runtime]
---

> The Mono C# VM is a rather nimble beast. Rather than requiring all

Reviewer: ```suggestion
The Mono .NET VM is a rather nimble beast. Rather than requiring all
```

> users live with the performance characteristics of a given policy, we

Reviewer: ```suggestion
users to live with the performance characteristics of a given policy, we
```

often choose to create multiple backends and bindings that exploit
what's best of the native platform while presenting a common
interface. Part of this is the choice between using an interpreter, a

> Just-In-Time compiler, or an Ahead-Of-Time compiler.

Reviewer: I think you need an introductory paragraph that actually introduces what we're doing (and the relative maturity of it:  is it an experiment, something we're going to ship soon, something that we shipped 12 months ago?) before diving into tradeoffs and details.

AOT compilation is attractive to some people for the
combination of optimized code with low start-up time. This is the
classic advantage of native code over code from a JIT or interpreter.
The performance of this code is often worse in raw throughput than can
be created by a JIT. This is because raw pointers to objects in memory
can't be placed into native code. Those pointers are unique to each
execution. When you're not generating code at runtime, you have to

> create wrappers for things to replace that level of indirection.

Reviewer: > create wrappers for things to replace that level of indirection.

Vague (and also wrong?). "wrappers for things" is really unclear.

Also "replace that level of indirection" doesn't make sense to me - the AOT compiler is the one that has to _introduce_ an indirection, because it cannot just refer directly to an object in memory - because it doesn't know where it would be.

The need to compensate for this performance loss can be more than
satisfied by the excellent LLVM code generator. This is not always
enough though. Sometimes, people just need more. Compiler optimization
is not an exact science. It has a lot of optimizations both in the big

> and the small.

Reviewer: > It has a lot of optimizations both in the big and the small.

Are you talking about interprocedural optimizations and peephole optimizations?  I'm not sure what this sentence is about.

Also I'm not sure what this paragraph is for.

I think this paragraph is supposed to reassure me that although AOT cannot play the same tricks as a JIT with directly addressing the actual objects that arise during the run of a program, LLVMs optimizations make up for it by generating better code than a JIT can since LLVM can spend more time analyzing the overall structure of the program.

Is that what you would like to get across?

> It can analyze your program globally and tell things that you could

Reviewer: ```suggestion
LLVM's optimization passes can analyze your program globally and tell things that you could
```

> never check yourself manually as a programmer before each commit.

Reviewer: In the previous paragraph you talk about "people", in this paragraph it is "you"/"yourself".  In the last couple of paragraphs it is "those people" and "they".  I would recommend to pick one.  "You" is probably the most casual tone; a generic "the programmer"/"they" would be closer to a more professional/academic tone.

> never check yourself manually as a programmer before each commit.

Reviewer: > tell things that you could never check yourself manually as a programmer before each commit.

Confusing why commits enter into it.  Also checking - I don't think of optimization passes as checking anything.

I think you want to say that a compiler is able to see through abstractions that programmers create in order to keep code manageable.  As programmers we use abstract data types and interfaces to separate concerns in our programs in order to make it easier for ourselves and our collaborators to reason about the behavior of our code.  We may use a function call to access a member of a data structure to maintain the illusion of abstraction because otherwise our code would be  a mess that is impossible to refactor or to explain.  A compiler on the other hand, is free to look through abstractions when generating code - it can inline accessor methods and discover recurring subexpressions, it can reorder data accesses or to replace expensive re-computations by incremental updates.

Likewise, it can examine the operations in a small segment of code and
make them perfect with respect to one another. Or it should be able
to. Sometimes the code we feed into LLVM does not match the patterns

> of inefficiency that it looks for, and we don't get an optimization.

Reviewer: "pattern of inefficiency" is an interesting term (and I actually found a couple of interesting papers that use that term), but  "optimization blocker" has way more relevant search hits.

Reviewer: On the other hand, "optimization blocker" makes it sound like Mono is actively butting heads with LLVM.  More likely we just don't annotate our LLVM IR in a way where LLVM can apply all of its optimization.

What's worse is that we don't know that we missed the optimization. We
don't know what we expect from code until it's a problem and we're

> really spending time optimizing it. Low-hanging fruit and dumper fires

Reviewer: ```suggestion
really spending time optimizing it. Low-hanging fruit and dumpster fires
```

This is really vague.  Maybe give some (brief) examples?

get priority.

> The solution to some manifestations of this problem is the use of

Reviewer: > some manifestations of this problem

vague and wishy-washy.  Also we're two paragraphs removed from the place where you briefly alluded to what the problem is (in the middle of a paragraph).

Say clearly what problem superoptimizers are solving for Mono.  Say it at the beginning of a paragraph.  Maybe add sections so there's a clear separation between the first section - where you're describing the performance difference between our AOT and JIT modes -  and the second section which is about the proposed solution.

superoptimizers. The academic discipline of superoptimizers is very
old. The idea is to treat the code that the programmer wrote as more
of a restriction, a specification. The superoptimizer generates a ton

> of randomly generated native code and checks how it behaves

Reviewer: ```suggestion
of random native code and checks whether it behaves
```

Reviewer: "generates a ton of randomly generated native code" is kinda clunky.  consider my suggested fix?
also `how`->`whether` is a grammar fix

differently than the written code. If it can generate a faster native

> code sequence than the compiler while keeping behavior exactly the

Reviewer: ```suggestion
code sequence than what the programmer wrote while keeping behavior exactly the
```

Reviewer: grammar fix.

Author: I think this is a bit misleading because the programmer didn't write the native code stream that's being compared against.  I can make that more clear though.

same, it wins.

This "exactly the same" part can be incredibly expensive if not done

> correctly. The computational effort involved has kept it from being

Reviewer: ```suggestion
correctly. The computational effort involved has kept superoptimization from being
```

Reviewer: Reconsider this?

> used very often. Computers get faster and theorems get better, and we

Reviewer: ```suggestion
used very often. As computers got faster, we
```

(I'm not sure what `and theorems got better` means.  Do you mean that we now have more accurate ways of specifying program behavior in a manner that can guide superoptimization in generating code?)

Reviewer: Still not a fan of "theorems get better".  what about "we learn more about compiler theory"?

are now seeing superoptimization research begin again.

One superoptimizer in particular, named Souper, has reached relative

> technical maturity while interoperating with the industry standard

Reviewer: > relative technical maturity

> industry standard

So we care because it's pretty stable and can easily plug into tools that we use?

code generator (LLVM) and the industry standard SMT engine (Z3). It
can take the LLVM IR that a compiler was going to feed into LLVM, and

> computes better IR. This can sometimes take a lot of time, and the

Reviewer: ```suggestion
compute better IR. This can sometimes take a lot of time, and the
```

Reviewer: Take this grammar fix?

code emitted is the result of a process that isn't auditable. You're
placing total faith in Souper for the correctness of your generated
code.

It's mostly useful for compiler engineers to use to tell that
optimizations were missed, and to identify how to fix that using

> conventional pattern matching. That said, Souper offers the ability to

Reviewer: > conventional pattern matching

That's... really not how a lot of high level optimizations work.  How about just "conventional techniques" since that will convey the point you want to make

Reviewer: I still don't like "conventional pattern matching" as a characterization of traditional compiler optimizations.

Author: I want to emphasize that they operate by matching a curated set of patterns (constant propagation, dead code elimination, loop unrolling, dataflow to eliminate dead stores) while souper just brute forces a search.

> drop in for clang and to generate the code that is run. Some folks

Reviewer: ```suggestion
drop in for LLVM's optimizer and backend and to emit the final superoptimized machine code. Some folks
```

like to live life in the fast lane (and assume the associated risks).
Some people may want to see just how fast mono could run their code if
the performance of generated code was not a factor (almost,
big-picture optimization still matters). Some people don't care if the

> compile time is longer.

Reviewer: This isn't the tone I'd use, but I'm not writing this blog.  If you want to keep it jokey and casual, maybe addressing "you" throughout the whole post is a good way to go.

Also maybe bullet points here?  People will skim long posts but pause over bulletpoints - it'll give them a place to anchor to see if this is relevant to what they want to do.

Something like:
```
   Superoptimization may be for you if:
   * you want to live life in the fast lane (and you trust machine code generated by Souper)
   * you want to see just how fast Mono could run your code if the performance of generated code was not a factor
     (but remember that writing efficient high-level code that uses performant algorithms still matters!)
   * you are willing to pay the penalty of longer compilation times.
```

For those people, I recommend they get a checkout of

[https://github.com/google/souper](https://github.com/google/souper)

and complete the compilation process described in that documentation.

When AOTing code with mono, they're going to want to pass the
commandline flags named there into the `---aot=llvmopts=` argument.

As of the time of this writing, that is

```
llvmopts="-load /path/to/libsouperPass.so -souper -z3-path=/usr/bin/z3" 
``` 

> Mono will then allow souper to step in during the middle of the LLVM

Reviewer: ```suggestion
Mono will then allow Souper to step in during the middle of the LLVM
```

compilation and try it's best at brute-forcing some better code. If
there's anything short and fast that does the job better, it will be

> found.

Reviewer: Add a conclusion paragraph.

Is this a call to action?  Is it something we want people to try and report back to us?   Is this something we did just to see if it's possible?  Does it demonstrate something about Mono's flexibility?  Are we abandoning the traditional LLVM backend?

Author: This does not supplant all of the optimization LLVM does, but operates as a pass on their IR. Superoptimization operates at the level of peephole optimization. I mentioned above that more big-picture (interprocedural) optimizations are still our responsibility and the responsibility of LLVM. I can be more clear about that.
